x-pack/plugins/security/public/management/api_keys/api_keys_grid/api_keys_grid_page.tsx

```diff
@@ -33,35 +22,108 @@ import { FormattedMessage } from '@kbn/i18n-react';
 import { APIKeysAPIClient } from '../api_keys_api_client';
 
+const parseSearchBarQuery = (query: Query): QueryContainer => {
+  let parsedQuery = query;
+
+  if (parsedQuery.hasSimpleFieldClause('owner')) {
+    const ownerQueryValue = parsedQuery.getSimpleFieldClause('owner')!.value;
+    parsedQuery = parsedQuery
+      .removeSimpleFieldClauses('owner')
+      .addSimpleFieldValue('username', `${ownerQueryValue}`);
+  }
+
+  if (query.text.includes('type:managed')) {
+    const subQuery = query.text.replace('type:managed', '');
+    parsedQuery = EuiSearchBar.Query.parse(`${subQuery} (metadata.managed:true OR Alerting*)`);
+  } else if (query.text.includes('type:rest') || query.text.includes('type:cross_cluster')) {
+    parsedQuery = EuiSearchBar.Query.parse(`${query.text} -metadata.managed:true -Alerting*`);
+  }
+
+  if (query.text.includes('expired:true')) {
+    const subQuery = query.text.replace('expired:true', '');
+    parsedQuery = EuiSearchBar.Query.parse(`${subQuery} expiration<=${moment.now()}`);
+  } else if (query.text.includes('expired:false')) {
+    const subQuery = query.text.replace('expired:false', '');
+    parsedQuery = EuiSearchBar.Query.parse(`${subQuery} expiration>${moment.now()}`);
+  }
+
+  parsedQuery = parsedQuery.addSimpleFieldValue('invalidated', false);
+
+  return EuiSearchBar.Query.toESQuery(parsedQuery);
+};
+
 export const APIKeysGridPage: FunctionComponent = () => {
   const { services } = useKibana<CoreStart>();
   const history = useHistory();
   const authc = useAuthentication();
-  const [state, getApiKeys] = useAsyncFn(
-    () => Promise.all([new APIKeysAPIClient(services.http).getApiKeys(), authc.getCurrentUser()]),
-    [services.http]
-  );
 
+  const [query, setQuery] = useState<Query>(EuiSearchBar.Query.parse(''));
+  const [from, setFrom] = useState<number>(0);
+  const [pageSize, setPageSize] = useState(25);
+  const [tableSort, setTableSort] = useState<QueryApiKeySortOptions>({
+    field: 'creation',
+    direction: 'desc',
+  });
   const [createdApiKey, setCreatedApiKey] = useState<CreateAPIKeyResult>();
   const [openedApiKey, setOpenedApiKey] = useState<CategorizedApiKey>();
   const readOnly = !useCapabilities('api_keys').save;
 
+  const [state, queryApiKeysAndAggregations] = useAsyncFn(() => {
+    const queryContainer = parseSearchBarQuery(query);
+
+    const requestBody = {
+      query: queryContainer,
+      from,
+      size: pageSize,
+      sort: tableSort,
+    };
+
+    return Promise.all([
+      new APIKeysAPIClient(services.http).queryApiKeys(requestBody),
+      authc.getCurrentUser(),
+    ]);
+  }, [services.http, query, from, pageSize, tableSort]);
+
+  const resetQueryOnError = () => {
+    setQuery(EuiSearchBar.Query.parse(''));
+    setFrom(0);
+    setTableSort({ field: 'creation', direction: 'desc' });
+  };
+
+  const onTableChange = ({ page, sort }: Criteria<ApiKey>) => {
+    setFrom(page?.index! * pageSize);
+    setPageSize(page?.size!);
```

Comment:
__question:__ I'm wondering when the `page` is not defined? It's a bit weird that we force it with `!`.

Reply:
From what i understand, `page` could be undefined and the `!` is just to satisfy the TS compiler. Even if `page` is present, the compiler thinks that `size` may not be guaranteed on that.

Reply:
Got it, thanks, weird types.